[quant][graph] Run dynamic quantization for specific ops

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37125 [quant][graph] Add JIT passes for dynamic quant multi uses of quant node
* **#37093 [quant][graph] Run dynamic quantization for specific ops**
* #37045 [graph][quant] Enable accessing child/grandchild modules in forward
* #37014 [quant][graph] Add check for qconfig_dict key

Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21283695](https://our.internmc.facebook.com/intern/diff/D21283695)

supriyar committed:
[quant][graph] Run dynamic quantization for specific ops
Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -270,13 +278,13 @@ bool mayRequireObservation(Value* v) {
 }
 
-bool nodeQuantizable(Node* n) {
+bool nodeQuantizable(Node* n, bool is_dynamic) {
```
you can have a default argument for is_dynamic?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1101,7 +1109,8 @@ bool InsertObserversHelper::valueNeedsToBeQuantized(Value* v) {
   if (!is_dynamic) {
     // Check whether producer is quantizable
-    if (mayRequireObservation(v) && nodeQuantizable(v->node())) {
+    if (mayRequireObservation(v) &&
+        nodeQuantizable(v->node(), /* is_dynamic */ false)) {
```
so that this can be reverted?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -53,6 +53,10 @@ std::vector<std::string> _quantizable_call_funcs = {
     "batch_norm",
 };
 
+std::vector<std::string> _quantizable_dynamic_call_funcs = {
```
maybe dynamic_quantizable?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -53,6 +53,10 @@ std::vector<std::string> _quantizable_call_funcs = {
+    "linear",
+};
+
 std::vector<std::string> _quantizable_aten_funcs = {
```
and you can rename this to static_quantizable as well

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1739,14 +1739,17 @@ def forward(self, x):
 
 
-    def test_insert_quant_dequant_conv_dynamic(self):
+    def test_dynamic_multi_op(self):
```
why is the name changed? this is supposed to be a unittest for insert_quant_dequant I think, you can add integration tests separately

supriyar replied:
There was already a separate test for `test_insert_quant_dequant_linear_dynamic` so this isn't necessary. I'll change it to an e2e test.

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1739,14 +1739,17 @@ def forward(self, x):
                 super(M, self).__init__()
-                self.conv = torch.nn.Conv2d(3, 5, 3).float()
+                self.conv = torch.nn.Conv2d(3, 1, kernel_size=3).to(dtype=torch.float)
+                self.fc1 = torch.nn.Linear(64, 10).to(dtype=torch.float)
 
             def forward(self, x):
-                return self.conv(x)
+                x = self.conv(x)
+                x = x.view(-1, 64).contiguous()
+                return self.fc1(x)
```
I think it's better to keep this as a unit test

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1872,6 +1826,25 @@ def forward(self, x):
         FileCheck().check("quantized::linear_dynamic") \
                    .run(model.graph)
 
+    def test_dynamic_multi_op(self):
```
nit: I think you can add a TestCase for integration tests for dynamic quant

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Changes looks good, I think we should add a test case for integrration tests for quantize_dynamic_script

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Run dynamic quantization for specific ops"


Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
